rivescript/rivescript.py

```diff
@@ -1658,6 +1671,62 @@ def current_user(self):
             self._warn("current_user() is meant to be used from within a Python object macro!")
         return self._current_user
 
+    def on(self, event_name, callback):
+        """Register an event callback.
+
+        Supported events (`event_name`):
+
+        * 'topic' - fired on topic change: `callback(user, topic)` or on 
+        redirect: `callback(user, topic, redirect=target)`
+
+        * 'uservar' - fired when user variable is set: callback(user, name, value)
+
+        Pass `None` as `callback` to clear it.
+        """
+        if not callback:
+            if event_name in self._callbacks:
+                self._say("Clearing callback for event '{}'".format(event_name))
+                del self._callbacks[event_name]
+            else:
+                self._warn("Callback for event '{}' is not set".format(event_name))
+        elif callable(callback):
+            self._say("Setting callback for event '{}'".format(event_name))
+            self._callbacks[event_name] = callback
+        else:
+            self._warn("Refusing to set callback for event '{}': {} is not callable".format(event_name, callback))
+
+    def get_topic(self, user):
+        """Get user's current topic.
+
+        :param str user: User name
+        :return: Topic or ``None`` if user does not exist.
+        """
+        if user in self._users:
+            return self._users[user]['topic']
+        else:
+            self._warn("Cannot get topic: user '{}' does not exist".format(user))
+            return None
+
+    def set_topic(self, user, topic):
+        """Set user's topic.
+
+        :param str user: User name
+        :param str topic: New topic
+        """
+        if topic in self._topics:
+            if not user in self._users:
+                self._users[user] = {}
+            self._say("Setting topic of user '{}' to '{}'".format(user, topic))
+            self._users[user]['topic'] = topic
+            self._fire_event('topic', user, topic)
+        else:
+            self._warn("Cannot set topic for user '{}': topic '{}' does not exist".format(user,topic))
+
+    def redirect(self, user, target):
+        """Redirect within current topic. Returns reply from the redirected trigger."""
+        return self._getreply(user, target)
```

Comment:
Why is this function here? To skip the processing of the `>begin` block? It doesn't seem to be called anywhere else internally except the unit tests. When redirects are actually processed it still goes straight for `self._getreply()`.